lib/Optimizer/Transforms/QuakeSynthesizer.cpp

```diff
@@ -115,6 +115,103 @@ static bool hasInitStateUse(BlockArgument argument) {
+
+  cudaq::IRBuilder irBuilder(builder);
+  irBuilder.genVectorOfConstants(argLoc, module, symbol, vec);
+
+  builder.setInsertionPointToStart(argument.getOwner());
+  return builder.create<cudaq::cc::AddressOfOp>(
+      argLoc, cudaq::cc::PointerType::get(arrTy), symbol);
+}
+
+template <typename T>
+LogicalResult synthesizeStateArgument(OpBuilder &builder, ModuleOp module,
+                                      unsigned &counter, BlockArgument argument,
+                                      Type eleTy, std::vector<T> &vec) {
+  auto *ctx = builder.getContext();
+  auto argLoc = argument.getLoc();
+  auto arrTy = cudaq::cc::ArrayType::get(ctx, eleTy, vec.size());
+
+  builder.setInsertionPointToStart(argument.getOwner());
+  auto toErase = std::vector<mlir::Operation *>();
+
+  // Iterate over the users of this state argument.
+  for (auto *argUser : argument.getUsers()) {
+    // Replace a calls to runtime function that reads the number of qubits
+    // with the log of the length, which is a synthesized constant.
+    if (auto numOfQubitsOp = dyn_cast<func::CallOp>(argUser)) {
+      if (auto calleeAttr = numOfQubitsOp.getCalleeAttr()) {
+        auto funcName = calleeAttr.getValue().str();
+        if (funcName == cudaq::getNumQubitsFromCudaqState) {
+          Value numOfQubits = builder.create<arith::ConstantIntOp>(
+              argLoc, log2(vec.size()), builder.getI64Type());
+          numOfQubitsOp.replaceAllUsesWith(ValueRange{numOfQubits});
+          toErase.push_back(numOfQubitsOp);
+        } else {
+          argUser->emitError("Unexpected call on state argument");
+          return failure();
+        }
+      }
+    }
+  }
+
+  auto buffer =
+      createGlobalArray(builder, module, counter, argument, arrTy, vec);
+  auto ptrArrEleTy =
+      cudaq::cc::PointerType::get(cudaq::cc::ArrayType::get(eleTy));
+  Value memArr = builder.create<cudaq::cc::CastOp>(argLoc, ptrArrEleTy, buffer);
+
+  argument.replaceAllUsesWith(memArr);
+
+  for (auto &op : toErase)
+    op->erase();
+
+  return success();
+}
+
+template <typename T>
+static LogicalResult synthesizeStateArgument(OpBuilder &builder,
+                                             ModuleOp module, unsigned &counter,
+                                             BlockArgument argument, Type eleTy,
+                                             void *data, std::size_t size) {
+  std::vector<T> vec;
+  for (std::size_t i = 0; i < size; i++) {
+    auto elePtr = reinterpret_cast<T *>(data) + i;
+    vec.push_back(*elePtr);
+  }
+  return synthesizeStateArgument(builder, module, counter, argument, eleTy,
+                                 vec);
+}
+
+static LogicalResult
+synthesizeStateArgument(OpBuilder &builder, ModuleOp module, unsigned &counter,
+                        BlockArgument argument,
+                        const cudaq::opt::ArgumentDataStore &stateData,
+                        std::size_t idx) {
```

Comment:
I thought this was all going to be done in a separate pass and not in QuakeSynthesis.

The synthesis of gates from the initial state value can and should be a distinct thing from argument synthesis. They are very different transformations in the IR. Having them separated will make it easy to use them and debug the passes when problems pop up.

It's super unfortunate that all these different transformation got labelled "synthesis". That may be the correct terminology in the literature, but it is rife with confusion and seems to suggest monolithic design instead of modularity perhaps.